service/pager/query_pagers.cc

```diff
@@ -355,21 +355,37 @@ class query_pager::query_result_visitor : public Base {
                     _rows_fetched_for_last_partition = v.last_partition_row_count;
                 }
             }
-            _last_pkey = v.last_pkey;
-            _last_ckey = v.last_ckey;
+            if (results->short_read_pos()) {
+                auto [last_pkey, last_ckey] = *results->short_read_pos();
+                _last_pkey = std::move(last_pkey);
+                _last_ckey = std::move(last_ckey);
+            } else {
+                _last_pkey = v.last_pkey;
+                _last_ckey = v.last_ckey;
+            }
         } else {
             row_count = results->row_count() ? *results->row_count() : std::get<1>(view.count_partitions_and_rows());
             _row_count = row_count;
             _max = _max - row_count;
             _exhausted = (row_count < page_size && !results->is_short_read()) || _max == 0;
 
-            if (!_exhausted && row_count > 0) {
+            if (!_exhausted) {
                 if (_last_pkey) {
                     update_slice(*_last_pkey);
                 }
-                auto [ last_pkey, last_ckey ] = view.get_last_partition_and_clustering_key();
-                _last_pkey = std::move(last_pkey);
-                _last_ckey = std::move(last_ckey);
+                if (results->short_read_pos()) {
+                    auto [last_pkey, last_ckey] = *results->short_read_pos();
+                    _last_pkey = std::move(last_pkey);
+                    _last_ckey = std::move(last_ckey);
```

Comment:
C++ question - doesn't the syntax
[_last_pkey, _last_ckey] = results->short_read_pos() work?

Reply:
No. auto [x, y] doesn't create new variables, it creates something new called bindings.

[existing_var_a, existing_var_b] don't work (but std::tie(a, b) = my_tuple does).

Reply:
If you have pre-existing variables you can use `std::tie()`:
```c++
std::tie(_last_pkey, _last_ckey) = std::move(*result->short_read_pos());
```